zetteldeft.el

```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
+      (insert-file-contents zdFile nil nil nil t)
+      (setq contents (concat (buffer-string))))
+    (setq title (deft-parse-title zdFile contents))
+    (kill-buffer "*Zetteldeft temp*")
+    title))
+
+(defun zetteldeft--rename (cusprompt fn-gen update-title)
+  "Generic function to rename a file.
+Takes CUSPROMPT, FN-GEN and UPDATE-TITLE to customise behaviour."
   (zetteldeft--check)
-    (let ((old-filename (buffer-file-name))
-          (deft-dir (file-name-as-directory deft-directory))
-          new-filename old-name new-name)
-      (when old-filename
-        (setq old-name (deft-base-filename old-filename))
-        (setq new-name (read-string
-                        (concat "Rename " old-name " to (without extension): ")
-                        old-name))
-        (setq new-filename
-              (concat deft-dir new-name "." deft-default-extension))
-        (rename-file old-filename new-filename)
-        (deft-update-visiting-buffers old-filename new-filename)
-        (zetteldeft-update-title-in-file)
-        (deft-refresh))))
-
-(defun zetteldeft-update-title-in-file ()
+  (let ((old-filename (buffer-file-name))
+	prompt-pars new-name new-filename)
+    (when old-filename
+      (setq prompt-pars (funcall cusprompt old-filename))
+      (setq new-name (read-string (car prompt-pars) (cdr prompt-pars)))
+      (setq new-filename (funcall fn-gen old-filename new-name))
+      (rename-file old-filename new-filename)
+      (deft-update-visiting-buffers old-filename new-filename)
+      (if update-title (zetteldeft-update-title-in-file new-name) nil)
+      (deft-refresh))))
+
+(defun zetteldeft-update-title-in-file (title)
   "Update the title of the current file, if present.
-Does so by looking for `zetteldeft-title-prefix'."
+Does so by looking for `zetteldeft-title-prefix'.
+Replaces current title with TITLE."
   (save-excursion
     (let ((zetteldeft-title-suffix ""))
       (goto-char (point-min))
       (when (re-search-forward (regexp-quote zetteldeft-title-prefix) nil t)
         (delete-region (line-beginning-position) (line-end-position))
-        (zetteldeft--insert-title)))))
+        (zetteldeft--insert-title title)))))
+
+(defun zetteldeft-file-rename ()
+  "Change current file's title, and use the new title to rename the file."
+  (interactive)
+  (zetteldeft--rename
+   (lambda (old-filename) (let (old-title prompt-text)
+			    (setq old-title (zetteldeft--lift-file-title old-filename))
+			    (setq prompt-text (concat "Change " old-title " to: "))
+			    `((,@prompt-text) (,@old-title))))
+   (lambda (old-filename new-name) (let (old-id old-extension new-filename)
+				     (setq old-id (zetteldeft--lift-id old-filename))
+				     (setq old-extension (zetteldeft--identify-extension old-filename))
+				     (setq new-filename (concat (file-name-sans-extension (deft-absolute-filename (concat old-id zetteldeft-id-filename-separator new-name))) "." old-extension))
```

Comment:
1. Creating unassigned variables and then calling `setq` is awkward. Prefer `let*`.
2. Indentation here doesn't keep with conventional elisp style.
3. Lines are quite long here, which gets hard to read.

Reply:
Thanks - updated. I hope the indentation and formatting changes have improved readability.